File: 18.4Sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        if(nums.size() < 4) return res;
        sort(nums.begin(), nums.end());
        int l,r,sum;
        int len = nums.size();
        for(int i=0; i< len-3; i++){
            if( i!=0 && nums[i] == nums[i-1])
                continue;
            for(int j = i+1; j<len-2; j++){
                if(j!=i+1 && nums[j] == nums[j-1])
                    continue;
                l = j+1;
                r = len-1;
                while(l < r){
                    sum = nums[i]+nums[j]+nums[l]+nums[r];
                    if( sum == target){
                        res.push_back({nums[i],nums[j],nums[l],nums[r]});
                        l++;
                        r--;
                        while( l< r && nums[l] == nums[l-1])
                            l++;
                        while( l< r && nums[r] == nums[r+1])
                            r--;
                    }
                    else if (sum < target)
                        l++;
                    else
                        r--;
                }
            }
        }
        return res;
    }
};

class Solution2 {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> rst;
        int len = nums.size(), sum;
        if(len < 4) return rst;
        int i, j, k, l;
        sort(nums.begin(), nums.end());
        for(i = 0; i < len - 3;){
            for(j = i + 1; j < len - 2;){
                k = j + 1;
                l = len - 1;
                while(k < l){
                    sum = nums[i] + nums[j] + nums[k] + nums[l];
                    if(sum == target){
                        rst.push_back({nums[i], nums[j], nums[k], nums[l]});
                        --l;
                        while(k < l && nums[l] == nums[l + 1]) --l;
                        ++k;
                        while(k < l && nums[k] == nums[k - 1]) ++k;
                    }
                    else if(sum < target){
                        ++k;
                        while(k < l && nums[k] == nums[k - 1]) ++k;
                    }
                    else{
                        --l;
                        while(k < l && nums[l] == nums[l + 1]) --l;
                    }
                }
                ++j;
                while(j < len - 2 && nums[j] == nums[j - 1]) ++j;
            }
            ++i;
            while(i < len - 3 && nums[i] == nums[i - 1]) ++i;
        }
        return rst;
    }
};

File: 18.4Sum.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = list(list())
        if len(nums)<4:
            return res
        nums.sort()
        n = len(nums)
        for i in range(0, n-3):
            if i!=0 and nums[i]==nums[i-1]:
                continue
            for j in range(i+1, n-2):
                if j!= i+1 and nums[j] == nums[j-1]:
                    continue
                l, r = j+1, n-1
                while l < r:
                    sum = nums[i] + nums[j] + nums[l] + nums[r];
                    if sum == target:
                        res.append([nums[i], nums[j], nums[l], nums[r]])
                        l, r = l+1, r-1
                        while l<r and nums[l]==nums[l-1]:
                            l = l + 1;
                        while l<r and nums[r]==nums[r+1]:
                            r = r - 1;
                    elif sum < target:
                        l = l + 1;
                    else:
                        r = r - 1;
        return res
                        
